File: Data Structures/Stack/stack.c
#include <stdio.h>
#include <stdlib.h>

//self referencing structure with one data element
struct Node {
    int data;
    struct Node *next;
};

// Aliasing for easier referencing
typedef struct Node Node; 

/*-------------------------------------- Stack Functions -------------------------------------- */
/** Ideally these should go in a seperate Header file (ie. stack.h) along with the self referencing structure
 * 
 * Stack Functions
 * Create Node   : create_node
 * Push          : push (adds an item to the top of the stack)
 * Pop           : pop (removes the topmost item from the stack)
 * Peek          : peek (take a look at the topmost item without popping)
 * Display Stack : print_stack
 * 
 */

Node * create_node(int data) {
    Node *new_node = malloc(1 * sizeof(Node));

    new_node->data = data;
    new_node->next = NULL;

    return new_node;
}


// adds a Node to the top of the stack
void push(Node **top, int data) {
    Node *new_node = create_node(data);

    if(*top != NULL) {
        new_node->next = *top;
    }

    *top = new_node;

}

// Pops the Top Node from the Stack
int pop(Node **top) {
    if(*top == NULL) {
        printf("Popping Failed: Empty Stack !!\n");
        return -1;
    }

    Node *popped = *top;

    *top = popped->next;

    int popped_int = popped->data;
    
    free(popped);

    printf("Successfully Popped : %d\n", popped_int);

    return popped_int;
}

// Prints all the elements in the Stack
void print_stack(Node *top) {
    if(top == NULL) {
        printf("Empty Stack!!\n");
        return;
    }

    printf("[");

    while(top != NULL) {
        printf(" %d", top->data);

        if(top->next != NULL) printf(",");

        top = top->next;
    }

    printf(" ]\n");
}

//Prints the topmost item in the Stack 
void peek(Node *top) {
    if(top == NULL) {
        printf("Empty Stack!!\n");
    }else {
        printf("First item of the Stack : %d\n", top->data);
    }
}

/* --------------------------------- End of Stack Functions ----------------------------------- */

int main() {
    Node *stack = NULL;

    push(&stack,10);
    push(&stack,20);
    push(&stack,50);

    pop(&stack);
    peek(stack);
    

    push(&stack,40);
    push(&stack,80);

    print_stack(stack);

}

File: Data Structures/Stack/Stack.java
class Stack {
    static final int MAX = 100; // Maximum size of Stack
    int top;
    int a[] = new int[MAX];
    Stack() { top = -1; }

    boolean push(int x) {
        if (top >= (MAX - 1)) {
            System.out.println("Stack Overflow");
            return false;
        }
        else {
            a[++top] = x;
            return true;
        }
    }

    int pop() {
        if (top < 0) {
            System.out.println("Stack Underflow");
            return 0;
        }
        else {
            int x = a[top--];
            return x;
        }
    }

    int peek() {
        if (top < 0) {
            System.out.println("Stack Underflow");
            return 0;
        }
        else {
            int x = a[top];
            return x;
        }
    }

    boolean isEmpty() {
        return (top < 0);
    }
}

// Driver code
class Main {
    public static void main(String args[])
    {
        Stack stk = new Stack();
        stk.push(1);
        stk.push(2);
        stk.push(3);
        System.out.println("1, 2, 3 added to stack");

        // Pop up Stack
        System.out.println(stk.pop() + " Popped from stack");

        // Get the top values
        System.out.println(stk.peek() + " is the peek value");

        // isEmpty
        if (stk.isEmpty()){
            System.out.println("Empty");
        }
        else {
            System.out.println("Not empty");
        }
    }
}